[Enhancement] Support MATCH expression for reading GIN

srlch opened the pull request:
## Why I'm doing:
In current implementation for GIN, if the row were tokenized, the behavior of the predicate would be inconsistent with the
original one. Users may be confused by this, and many optimizer rewriting rules may conflict with this behavior change, which may easily lead to incorrect query results.

## What I'm doing:
In this pr, we introduce a NEW expression, says `(NOT) MATCH` expression. We use `(NOT) MATCH` expression to use the GIN to filter the data. The specific behavior is as follows:

For the parser type is `none`:
`(NOT) MATCH '%abc%'` have exactly the same behavior as `LIKE`. It should be noted that when parser is `none`, all the following predicate can use GIN to accelerate the filtering without any behavior changing: `<=, >=, ==, LIKE, IS NULL/IS NOT NULL, IN, NOT IN`.

For the parser type is `english/chinese/standard`:
`(NOT) MATCH '%abc%'` means that match the keyword after the row has been tokenized. In this case, only using `MATCH` expression can filter the data using GIN.

Pay attention:
1. `MATCH` expression can be only used for column with GIN.
2. `MATCH` expression can only be used as a pushdown predicate. It means that we can not call this expression directly

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

packy92 commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -954,6 +958,57 @@ public Void visitLikePredicate(LikePredicate node, Scope scope) {
+                throw new SemanticException("left operand of MATCH must be column ref");
+            }
+
+            for (Field f : scope.getRelationFields().getAllFields()) {
```
please add some tests for query like `select * from (select t1.cola as colb from t1 join t2) t1 where t1.colb match 'abc'`. Is the query like this can use GIN?

srlch replied:
Updated, remove the check code here. BE will return error if the column using match has no gin

srlch committed:
[Enhancement] Support MATCH expression for reading GIN

Why I'm doing:
In current implementation for GIN, if the row were tokenized,
the behavior of the predicate would be inconsistent with the
original one. Users may be confused by this, and many optimizer
rewriting rules may conflict with this behavior change, which
may easily lead to incorrect query results.

What I'm doing:
In this pr, we introduce a NEW expression, says '(NOT) MATCH'
expression. We use (NOT) MATCH expression to use the GIN to
filter the data. The specific behavior is as follows:

For the parser type is 'none':
(NOT) MATCH '%abc%' have exactly the same behavior as LIKE.
It should be noted that when parser is 'none', all the following
predicate can use GIN to accelerate the filtering without any behavior
changing: <=, >=, ==, LIKE, IS NULL/IS NOT NULL, IN, NOT IN

For the parser type is 'english'/'chinese'/'standard':
(NOT) MATCH '%abc%' means that match the keyword after the row has been
tokenized. In this case, only using MATCH expression can filter the data
using GIN.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix sql test

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
add ut

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix

Signed-off-by: srlch <linzichao@starrocks.com>

packy92 commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -954,6 +956,26 @@ public Void visitLikePredicate(LikePredicate node, Scope scope) {
+                throw new SemanticException("left operand of MATCH must be column ref");
+            }
+
+            if (!type2.isStringType() && !type2.isNull()) {
```
Is the sceond parameter should be a constant string value?

srlch replied:
I think string type is fine, if the second parameter is a function with return type as string, it is also avaliable.

wyb commented on fe/fe-core/src/main/java/com/starrocks/analysis/MatchExpr.java:
```diff
@@ -0,0 +1,83 @@
+        return new MatchExpr(this);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!super.equals(obj)) {
+            return false;
+        }
+        return true;
+    }
```
not necessary

srlch replied:
done

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:
```diff
@@ -2813,7 +2814,7 @@ nonReserved
     | JOB
     | LABEL | LAST | LESS | LEVEL | LIST | LOCAL | LOCATION | LOGS | LOGICAL | LOW_PRIORITY | LOCK | LOCATIONS
-    | MASKING | MANUAL | MAP | MAPPING | MAPPINGS | MATERIALIZED | MAX | META | MIN | MINUTE | MODE | MODIFY | MONTH | MERGE | MINUS
+    | MATCH | MASKING | MANUAL | MAP | MAPPING | MAPPINGS | MATERIALIZED | MAX | META | MIN | MINUTE | MODE | MODIFY | MONTH | MERGE | MINUS
```
```suggestion
    | MANUAL | MAP | MAPPING | MAPPINGS | MASKING | MATCH | MATERIALIZED | MAX | META | MIN | MINUTE | MODE | MODIFY | MONTH | MERGE | MINUS
```

srlch replied:
done

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/plan/ScalarOperatorToExpr.java:
```diff
@@ -378,6 +380,14 @@ public Expr visitLikePredicateOperator(LikePredicateOperator predicate, Formatte
+        public Expr visitMatchExprOperator(MatchExprOperator operator, FormatterContext context) {
+            Expr child1 = buildExpr.build(operator.getChild(0), context);
+            Expr child2 = buildExpr.build(operator.getChild(1), context);
+            MatchExpr expr = new MatchExpr(child1, child2);
+            return expr;
```
```suggestion
            return new MatchExpr(child1, child2);
```

srlch replied:
done

wyb commented on fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/MatchExprOperator.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.List;
+import java.util.Objects;
+
+public class MatchExprOperator extends ScalarOperator {
```
why not extends `PredicateOperator`

srlch replied:
It will be simpler to implement MATCH defined as a normal expression than a predicate on the BE side.

wyb commented on fe/fe-core/src/main/java/com/starrocks/analysis/MatchExpr.java:
```diff
@@ -0,0 +1,83 @@
+
+import java.util.Objects;
+
+public class MatchExpr extends Expr {
```
why not extends `Predicate`

srlch replied:
same reason

wyb commented on be/src/storage/column_expr_predicate.cpp:
```diff
@@ -285,32 +285,62 @@ Status ColumnExprPredicate::try_to_rewrite_for_zone_map_filter(starrocks::Object
+                 expr->get_num_children() == 2 && expr->get_child(0)->node_type() == TExprNodeType::SLOT_REF;
+
+    // check if satisfy vaild MATCH format
+    vaild_match = vectorized_function_call == nullptr && expr->node_type() == TExprNodeType::MATCH_EXPR &&
```
`match` is not function?

srlch replied:
I have defined it as MATCH_EXPR which is a new type

srlch committed:
Fix comment

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix

Signed-off-by: srlch <linzichao@starrocks.com>
